implementations/rust/ockam/ockam_api/src/kafka/secure_channel_map.rs

```diff
@@ -0,0 +1,387 @@
+
+                let message = SecureChannelIdentifierMessage {
+                    secure_channel_identifier: random_unique_id,
+                };
+
+                //communicate to the other end the random id associated with this
+                //secure channel, and wait to an empty reply to avoid race conditions
+                //on the order of encryption/decryption of messages
+                context
+                    .send_and_receive(
+                        route![
+                            encryptor_address.clone(),
+                            KAFKA_SECURE_CHANNEL_CONTROLLER_ADDRESS
+                        ],
+                        message,
+                    )
+                    .await?;
+
+                trace!("assigned id {random_unique_id} to {topic_partition_address}");
+                (random_unique_id, encryptor_address)
+            }
+        };
+
+        inner
+            .identity
+            .secure_channel_registry()
+            .get_channel_by_encryptor_address(&encryptor_address)
+            .map(|entry| (random_unique_id, entry))
+            .ok_or_else(|| Error::new(Origin::Channel, Kind::Unknown, "secure channel down"))
+    }
+
+    ///return decryptor api address
+    async fn get_secure_channel_for(
+        &self,
+        secure_channel_id: UniqueSecureChannelId,
+    ) -> Result<SecureChannelRegistryEntry, Error> {
+        let inner = self.inner.lock().await;
+        if let Some(encryptor_address) = inner.id_encryptor_map.get(&secure_channel_id) {
+            inner
+                .identity
+                .secure_channel_registry()
+                .get_channel_list()
+                .iter()
+                .find(|entry| {
+                    entry.encryptor_messaging_address() == encryptor_address
+                        && !entry.is_initiator()
+                })
+                .cloned()
+                .ok_or_else(|| {
+                    Error::new(
+                        Origin::Channel,
+                        Kind::Unknown,
+                        "secure channel no longer exists",
+                    )
+                })
+        } else {
+            Err(Error::new(
+                Origin::Channel,
+                Kind::Unknown,
+                "missing secure channel",
+            ))
+        }
+    }
+}
+
+#[async_trait]
+impl<V: IdentityVault, S: AuthenticatedStorage> KafkaSecureChannelController
+    for KafkaSecureChannelControllerImpl<V, S>
+{
+    async fn encrypt_content_for(
+        &self,
+        context: &mut Context,
+        topic_name: &str,
+        partition_id: i32,
+        content: Vec<u8>,
+    ) -> Result<(UniqueSecureChannelId, Vec<u8>), Error> {
+        let (unique_id, secure_channel_entry) = self
+            .get_or_create_secure_channel_for(context, topic_name, partition_id)
+            .await?;
+
+        trace!("encrypting content with {unique_id}");
+        let encryption_response: EncryptionResponse = context
```

Comment:
I think that's the idea in the long run

Reply:
@etorreborre that would be great, unfortunately, that's non-trivial to implement, because we can't send unserializable structs over the messaging system

Reply:
@SanjoDeundiak I would like to understand this constraint better because I don't see why the code for encrypting messages can not be in principle called directly from the `encrypt_content_for` function. Let's say that you kept a pointer to the encoder of a secure channel in the secure channel registry and called the encoder directly without messaging it. Would that work?